_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+    Product product = new Product(1L, "Awesome Jeans",
+        BigDecimal.valueOf(49.99), BigDecimal.valueOf(25.00));
+    productEventHandler.handleCreated(product);
+  }
+}
+```
+
+With the contract and test base class given above, a test class `ContractVerifierTest` is generated into `build/generated-test-sources/contracts/org/springframework/cloud/contract/verifier/tests/`.
+The test is calling the trigger method stated in the contract and then captures the event using `ContractVerifierMessaging`.
+The way an event is captured using plain Spring Amqp is part of the Amqp integration in Spring Cloud Contract that I contributed a while ago.
+It is achieved by spying on the `RabbitTemplate`.
+The test then runs the assertions from the `outputMessage` section of our contract against the event.
+
+```java
+public class ContractVerifierTest extends ProductPublisherTestBase {
+
+  @Inject ContractVerifierMessaging contractVerifierMessaging;
+  @Inject ContractVerifierObjectMapper contractVerifierObjectMapper;
+
+  @Test
+  public void validate_shouldProduceValidProdcutEvent() throws Exception {
+    // when:
+    emitProductCreatedEvent();
+
+    // then:
+    ContractVerifierMessage response = contractVerifierMessaging.receive("test-exchange");
+    assertThat(response).isNotNull();
+    assertThat(response.getHeader("contentType")).isEqualTo("application/json");
+    // and:
+    DocumentContext parsedJson = JsonPath.parse(contractVerifierObjectMapper.writeValueAsString(response.getPayload()));
+    assertThatJson(parsedJson).field("name").isEqualTo("Awesome Jeans");
+    assertThatJson(parsedJson).field("purchasePrice").isEqualTo(25.00);
+    assertThatJson(parsedJson).field("id").matches("[0-9]+");
+    assertThatJson(parsedJson).field("salesPrice").isEqualTo(49.99);
+  }
+}
+```
+
+So if we would change an attribute name in `Product` we would have a failing test.
+**This shields us from making accidental incompatible contract-relevant changes**.
+
+The contracts of the product service are published in a stub jar to a maven repository.
+
+This stub jar can be used by the checkout service.
+This is done by stating the stub information in the [application configuration](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/resources/application.yml) of the checkout service.
+
+```yml
+stubrunner:
+  work-offline: true
+  ids: 'com.epages:product-service'
+```
+
+Now all we need to do is annotate our improved test on checkout side with `@AutoConfigureStubRunner` and use the `StubTrigger` to send the stub message in.
+The `spring-cloud-contract-stub-runner` downloads the stub jar with our contract from a maven repository and emits it into our application.
+This is done without interaction with a running broker.
+After the event is emitted we can run our assertions to make sure the correct effect can be observed.
+In our case, we assert that the correct product reference has been saved.
```

Comment:
assert is a term from the test framework we use - an assertion is used to check a post condition....